### PR TITLE
Barcode

### PR DESCRIPTION
**Pull request by flisowna**
## Change Summary

Provide summary of changes with issue number if any.

## Change type

- [ ] feat: (new feature for the user, not a new feature for build script)
- [ ] fix: (bug fix for the user, not a fix to a build script)
- [ ] docs: (changes to the documentation)
- [ ] style: (formatting, missing semi colons, etc; no production code change)
- [ ] refactor: (refactoring production code, eg. renaming a variable)
- [ ] test: (adding missing tests, refactoring tests; no production code change)
- [ ] chore: (updating grunt tasks etc; no production code change)

## Test/ Verification

Provide summary of changes.

## Additional information / screenshots (optional)

Anything for maintainers to be made aware of

### CODE REVIEW

File: packages/nc-gui/components/smartsheet/column/BarcodeOptions.vue [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+onMounted(() => {
+  // set default value
+  vModel.value.fk_barcode_value_column_id = (column?.value?.colOptions as Record<string, any>)?.fk_barcode_value_column_id || ''
+})
+
+setAdditionalValidations({
+  fk_barcode_value_column_id: [{ required: true, message: 'Required' }],
+  barcode_format: [{ required: true, message: 'Required' }],
+})
```
Comment (wingkwong) [inline packages/nc-gui/components/smartsheet/column/BarcodeOptions.vue:61]: Need a better message

image

File: packages/nc-gui/components/smartsheet/Gallery.vue [hunk 2]
```diff
@@ -1,5 +1,6 @@
 <script lang="ts" setup>
 import { ViewTypes, isVirtualCol } from 'nocodb-sdk'
+import VueBarcode from '@chenfengyuan/vue-barcode'
```
Comment (wingkwong) [inline packages/nc-gui/components/smartsheet/Gallery.vue:3]: Redundant? Ignore this comment if this is still wip.

File: packages/nc-gui/plugins/ant.ts [hunk 3]
```diff
@@ -1,7 +1,9 @@
 import { Menu as AntMenu, Modal as AntModal, message } from 'ant-design-vue'
+// import VueBarcode from '@chenfengyuan/vue-barcode'
```
Comment (wingkwong) [inline packages/nc-gui/plugins/ant.ts:2]: Redundant? Ignore this comment if this is still wip.

File: packages/nc-gui/components/virtual-cell/barcode/JsBarcodeWrapper.vue [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+const onBarcodeClick = (ev: MouseEvent) => {
+  ev.stopPropagation()
+  emit('onClickBarcode')
+}
+
+watch(() => props.barcodeValue, generate)
+watch(() => props.barcodeFormat, generate)
+onMounted(generate)
```
Comment (wingkwong) [inline packages/nc-gui/components/virtual-cell/barcode/JsBarcodeWrapper.vue:30]: some line breaks would be better

File: packages/nc-gui/components/virtual-cell/barcode/JsBarcodeWrapper.vue [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+<template>
+  <svg data-testid="barcode" v-show="!errorForCurrentInput" ref="barcodeSvgRef" @click="onBarcodeClick"></svg>
+  <slot v-if="errorForCurrentInput" name="barcodeRenderError" />
+</template>
+
+<style scoped>
+svg {
+  width: 100%;
+}
+</style>
```
Comment (wingkwong) [inline packages/nc-gui/components/virtual-cell/barcode/JsBarcodeWrapper.vue:42]: can be moved to class.

```js
<svg class="w-full" .../>
```

File: packages/nocodb/src/lib/migrations/XcMigrationSourcev2.ts [hunk 6]
```diff
@@ -10,6 +10,7 @@ import * as nc_019_add_meta_in_meta_tables from './v2/nc_019_add_meta_in_meta_ta
 import * as nc_020_kanban_view from './v2/nc_020_kanban_view';
 import * as nc_021_add_fields_in_token from './v2/nc_021_add_fields_in_token';
 import * as nc_022_qr_code_column_type from './v2/nc_022_qr_code_column_type';
+import * as nc_023_barcode_column_type from './v2/nc_023_barcode_column_type';
```
Comment (wingkwong) [inline packages/nocodb/src/lib/migrations/XcMigrationSourcev2.ts:13]: 023 has been used. You may update it when everything is reviewed and ready for merge.

File: packages/nocodb/src/lib/migrations/v2/nc_023_barcode_column_type.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    table.foreign('fk_column_id').references(`${MetaTable.COLUMNS}.id`);
+
+    table.string('fk_barcode_value_column_id', 20);
+    table
+      .foreign('fk_barcode_value_column_id')
+      .references(`${MetaTable.COLUMNS}.id`);
+
+    table.string('barcode_format', 15);
+    table.boolean('deleted');
+    table.float('order');
```
Comment (wingkwong) [inline packages/nocodb/src/lib/migrations/v2/nc_023_barcode_column_type.ts:18]: What does `order` do for barcode?

File: packages/nocodb/src/lib/models/BarcodeColumn.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        MetaTable.COL_BARCODE,
+        { fk_column_id: columnId }
+      );
+      await NocoCache.set(`${CacheScope.COL_BARCODE}:${columnId}`, column);
+    }
+
+    return column ? new BarcodeColumn(column) : null;
+  }
+
+  id: string;
```
Comment (wingkwong) [inline packages/nocodb/src/lib/models/BarcodeColumn.ts:47]: better put it back to the top (~ Line 7)

**Review by wingkwong — COMMENTED** [review]
I think it'd be better to preset the column 

image

File: packages/nc-gui/components/smartsheet/column/BarcodeOptions.vue (hunk 1 above)
Comment (flisowna) [reply packages/nc-gui/components/smartsheet/column/BarcodeOptions.vue:61]: I preset the field so the problem arises now only when there's not appropriate field yet. Here how it looks like: 


Screenshot 2022-12-21 at 13 04 35

Do you think it's correct way to resolve this issue?

File: packages/nocodb/src/lib/migrations/XcMigrationSourcev2.ts (hunk 6 above)
Comment (flisowna) [reply packages/nocodb/src/lib/migrations/XcMigrationSourcev2.ts:13]: Yes I'm aware of it:) thank you

File: packages/nc-gui/components/smartsheet/column/BarcodeOptions.vue (hunk 1 above)
Comment (wingkwong) [reply packages/nc-gui/components/smartsheet/column/BarcodeOptions.vue:61]: What about adding a small question mark icon next to `Column with Barcode value` as a help text? Like hovering would show the available column types for barcode type. In this way, we can keep the no-data-text short (we can say no valid column type, please refer to the help text blah blah blah) and at the same time we can still allow users to check that even if there are some available columns.

Comment (flisowna) [reply packages/nc-gui/components/smartsheet/column/BarcodeOptions.vue:61]: Yey! that's sound very good. Thank you!

Comment (flisowna) [reply packages/nc-gui/components/smartsheet/column/BarcodeOptions.vue:72]: I've changed it, can you check it?

Comment (wingkwong) [reply packages/nc-gui/components/smartsheet/column/BarcodeOptions.vue:72]: It looks good!

**Review by wingkwong — APPROVED** [review]

File: packages/nc-gui/components/virtual-cell/barcode/Barcode.vue [hunk 9]
```diff
@@ -26,6 +26,11 @@ const barcodeMeta = $computed(() => {
 
 const handleModalOkClick = () => (modalVisible.value = false)
 
+// const showBarcode = computed(() => barcodeValue?.value?.length > 0 && !tooManyCharsForBarcode.value)
+const showBarcode = computed(() => {
+  return cellValue?.value?.length > 0 && !tooManyCharsForBarcode.value
+})
```
Comment (wingkwong) [inline packages/nc-gui/components/virtual-cell/barcode/Barcode.vue:31]: why not just 

```js
const showBarcode = computed(() => barcodeValue?.value?.length > 0 && !tooManyCharsForBarcode.value)
```

File: packages/nc-gui/components.d.ts [hunk 10]
```diff
@@ -201,6 +201,7 @@ declare module '@vue/runtime-core' {
     MdiLogin: typeof import('~icons/mdi/login')['default']
     MdiLogout: typeof import('~icons/mdi/logout')['default']
     MdiMagnify: typeof import('~icons/mdi/magnify')['default']
+    MdiMapMarker: typeof import('~icons/mdi/map-marker')['default']
```
Comment (wingkwong) [inline packages/nc-gui/components.d.ts:204]: i don't see MdiMapMarker is used in this PR

File: packages/nc-gui/components/smartsheet/column/BarcodeOptions.vue [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+
+const showBarcodeValueColumnInfoIcon = computed(() => !columnsAllowedAsBarcodeValue.value?.length)
+
+// set default meta value
```
Comment (wingkwong) [inline packages/nc-gui/components/smartsheet/column/BarcodeOptions.vue:76]: this comment seems redundant.

File: packages/nc-gui/components/virtual-cell/QrCode.vue [hunk 12]
```diff
@@ -6,8 +6,8 @@ const maxNumberOfAllowedCharsForQrValue = 2000
 const cellValue = inject(CellValueInj)
 
 const qrValue = computed(() => String(cellValue?.value))
-
 const tooManyCharsForQrCode = computed(() => qrValue?.value.length > maxNumberOfAllowedCharsForQrValue)
+const showQrCode = computed(() => qrValue?.value?.length > 0 && !tooManyCharsForQrCode.value)
```
Comment (wingkwong) [inline packages/nc-gui/components/virtual-cell/QrCode.vue:10]: - add a new line between L8 and L9.
- add a new line between L9 and L10.

File: packages/nc-gui/components/virtual-cell/barcode/Barcode.vue [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+<template>
+  <a-modal
+    v-model:visible="modalVisible"
+    :class="{ active: modalVisible }"
+    wrap-class-name="nc-barcode-large"
+    :body-style="{ padding: '0px' }"
+    :footer="null"
+    @ok="handleModalOkClick"
+  >
+    <JsBarcodeWrapper v-if="showBarcode" :barcode-value="barcodeValue" :barcode-format="barcodeMeta.barcodeFormat">
```
Comment (wingkwong) [inline packages/nc-gui/components/virtual-cell/barcode/Barcode.vue:45]:
```ts
<JsBarcodeWrapper v-if="showBarcode" :barcode-value="barcodeValue" :barcode-format="barcodeMeta.barcodeFormat" />
```

File: packages/nc-gui/components/virtual-cell/barcode/JsBarcodeWrapper.vue [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+import JsBarcode from 'jsbarcode'
+
+const props = defineProps({
+  barcodeValue: { type: String, required: true },
+  barcodeFormat: { type: String, required: true },
+})
+
+const emit = defineEmits(['onClickBarcode'])
+const barcodeSvgRef = ref(null)
+const errorForCurrentInput = ref(false)
```
Comment (wingkwong) [inline packages/nc-gui/components/virtual-cell/barcode/JsBarcodeWrapper.vue:11]: some line breaks

File: packages/nc-gui/components/virtual-cell/barcode/JsBarcodeWrapper.vue [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+  }
+}
+
+const onBarcodeClick = (ev: MouseEvent) => {
+  ev.stopPropagation()
+  emit('onClickBarcode')
+}
+
+watch(() => props.barcodeValue, generate)
+watch(() => props.barcodeFormat, generate)
```
Comment (wingkwong) [inline packages/nc-gui/components/virtual-cell/barcode/JsBarcodeWrapper.vue:31]: watch API can take multiple sources with the single handler. You can combine them to

```ts
watch([() => props.barcodeValue, () => props.barcodeFormat], generate)
```

File: packages/nc-gui/components/virtual-cell/barcode/JsBarcodeWrapper.vue [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+onMounted(generate)
```
Comment (wingkwong) [inline packages/nc-gui/components/virtual-cell/barcode/JsBarcodeWrapper.vue:32]: add below line at the top

```ts
import { onMounted } from '#imports'
```

File: packages/nc-gui/plugins/ant.ts (hunk 3 above)
Comment (wingkwong) [inline packages/nc-gui/plugins/ant.ts:2]: can be removed if this is not required

File: packages/nc-gui/plugins/ant.ts [hunk 17]
```diff
@@ -1,7 +1,9 @@
...
 import { defineNuxtPlugin } from '#imports'
 
 export default defineNuxtPlugin((nuxtApp) => {
+  // nuxtApp.vueApp.component(VueBarcode.name, VueBarcode)
```
Comment (wingkwong) [inline packages/nc-gui/plugins/ant.ts:6]: can be removed if this is not required

**Review by wingkwong — CHANGES_REQUESTED** [review]
After the changes, you may run `npm run lint` to lint the code and fix the corresponding lint issues if applicable.

File: packages/nc-gui/components/virtual-cell/barcode/JsBarcodeWrapper.vue (hunk 15 above)
Comment (flisowna) [reply packages/nc-gui/components/virtual-cell/barcode/JsBarcodeWrapper.vue:31]: Thank you, it's a good hint:)

File: packages/nc-gui/components/virtual-cell/barcode/Barcode.vue (hunk 9 above)
Comment (flisowna) [reply packages/nc-gui/components/virtual-cell/barcode/Barcode.vue]: It was a bit trickier as I thought. Please check if it's ok now.

Comment (wingkwong) [reply packages/nc-gui/components/virtual-cell/barcode/Barcode.vue]: Now It's fine. Basically they are same. Just to keep it clean as `return` is implied in this way.

**Comment by wingkwong** [comment]
@flisowna is it ready for the final review?

**Comment by flisowna** [comment]
@wingkwong thank you :) 

> After the changes, you may run `npm run lint` to lint the code and fix the corresponding lint issues if applicable.

I think I fixed now all lint issues of the files that I touched (the linter complaint about more files though, so also ones that I didn't change). 
Also, I think the linter command is only configured at the moment for the FE and the BE packages, but not the SDK. 

If you agree that this is sufficient for the scope of this PR, from my side it's ready for the final review and merge :)

**Review by wingkwong — APPROVED** [review]

**Pull request merged**